extensions/amp-timeago/0.2/use-intersect.js

```diff
@@ -0,0 +1,40 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {useLayoutEffect, useRef, useState} from '../../../src/preact';
+
+/**
+ * @param {object} ref
+ * @return {?Array<IntersectionObserverEntry>}
+ */
+export function useIntersect(ref) {
+  const {0: entries, 1: set} = useState([]);
+  const observerRef = useRef(null);
+  if (observerRef.current === null) {
+    observerRef.current = new IntersectionObserver(set);
+  }
+
+  useLayoutEffect(() => {
+    const {current: node} = ref;
+    const {current: observer} = observerRef;
+    observer.disconnect();
```

Comment:
According to the React [docs](https://reactjs.org/docs/hooks-effect.html#example-using-hooks-1)

>  effects run for every render and not just once. This is why React also cleans up effects from the previous render before running the effects next time

Do we still need this `disconnect` / whats it for?

Reply:
IIUC the first `disconnect` is to stop observing prior nodes observed by the `observerRef` when `ref` changes, since we only want to observe one at a time in this case. The callback `disconnect` is to stop observing any nodes observed by the state of the `observerRef` at the time the effect was defined. This accounts for the case where `observerRef` changes, so we can still unmount for props we no longer have access to. [This is a helpful resource to how the `useEffect` hook handles cleanup.](https://overreacted.io/a-complete-guide-to-useeffect/#so-what-about-cleanup)

Reply:
Discussed offline -- we don't appear to need the outer disconnect as it is redundant with the callback based on manual testing. @jridgewell to weigh in here